lib/HLSL/HLOperationLower.cpp

```diff
@@ -4421,6 +4421,40 @@ static Value* SkipAddrSpaceCast(Value* Ptr) {
   return Ptr;
 }
 
+// For known non-groupshared, verify that the destination param is valid
+void ValidateAtomicDestination(CallInst *CI, HLObjectOperationLowerHelper *pObjHelper) {
+  Value *dest = CI->getArgOperand(HLOperandIndex::kInterlockedDestOpIndex);
+  GetElementPtrInst *gep = nullptr;
+
+  // If we encounter a gep, we may provide a more specific error message
+  while (isa<GetElementPtrInst>(dest)) {
+    gep = cast<GetElementPtrInst>(dest);
+    dest = gep->getPointerOperand();
+  }
+
+  // Confirm that dest is a UAV
+  if (CallInst *destCall = dyn_cast<CallInst>(dest)) {
+    hlsl::HLOpcodeGroup group = hlsl::GetHLOpcodeGroup(destCall->getCalledFunction());
+    while (group == HLOpcodeGroup::HLSubscript) {
+      dest = destCall->getArgOperand(HLOperandIndex::kSubscriptObjectOpIdx);
+      if (!isa<CallInst>(dest))
+        break;
+      destCall = cast<CallInst>(dest);
+      group = hlsl::GetHLOpcodeGroup(destCall->getCalledFunction());
+    }
+    DXIL::ResourceKind RK = pObjHelper->GetRK(destCall);
+    if (DXIL::IsStructuredBuffer(RK))
```

Comment:
Shouldn't this also fail if not UAV?  Or was that caught elsewhere?  Perhaps while translating the subscript operation?

Reply:
Non-UAV resources are considered const at codegen and able to be caught there. Non-resource parameters will fail all these checks and hit the catchall at the end of this function

Reply:
We have a property right here identifying this as a UAV or not.  Perhaps we should at least DXASSERT that it is a UAV, if the error should normally be caught elsewhere (with comment stating that to explain why it's just an assert here).

Reply:
The DXASSERT would have to be pretty complicated because while SRVs can't be here, the resource could be invalid because it's not actually a resource or possibly other values. I'll just add a mostly redundant check